### ISSUE TITLE
Matching collections

### ISSUE THREAD

**Issue by elgerm** [NRules/NRules#42]
Hi, very nice library.

Probably not an issue, also posted this on [stackoverflow](http://stackoverflow.com/questions/28190119/nrules-match-a-collection): how to match a collection instead of a single variable?

Something like this:
'''
IEnumerable<Order> orders = null;

When()
    .Match<IEnumerable<Order>>(o => o.Where(c => c.Cancelled).Count() >= 3)
        .Collect<Order>(() => orders, o => o.Cancelled);

Then().Do(ctx => orders.ToList().ForEach(o => o.DoSomething()));
'''
never matches (although i insert more then 3 cancelled orders), but a single cancelled order (ie declare a single order in a rule) does match, can you please explain why? 

Thanks a lot.

**Comment by elgerm** [comment]
Nevermind... figured it out, ofcourse you have to insert a collection. Maybe in the help docs a bit about how the local variables are inserted. I'd be willing to contribute to the docs.

**Comment by gimmemoore** [comment]
I think this might be an issue though. When using only the property as in 

Collect(() => orders, o => **o.Cancelled**);

NRules somehow fails to evaluate the condition to true, but if you put **o.Cancelled == true** it works because this now becomes an expression supported.

**Comment by snikolayev** [comment]
@elgerm Regarding matching a collection. You probably don't want to do that (i.e. insert the whole collection), instead, you would use Collect. Though, right now Collect does not allow you to specify conditions on the resulting aggregation - I was thinking about enhancing the engine to do that.
I'll see if I can clarify in the docs.

**Comment by snikolayev** [comment]
@gimmemoore Regarding matching on a property - I'll take a look

**Comment by snikolayev** [comment]
Kevin, I tried matching on a Boolean property directly the way you described and it works fine. Which version are you using (I tried 0.2.4 from the nuget feed)? Also, which .NET runtime (I tried, 4.0, 4.5 and 4.5.1)? Can you post a simple example with the domain model definition and a rule that exhibits the behavior?

Here is my example that works fine (can you please try it and see if it works?):
```c#
    public class Order
    {
        public bool Cancelled { get; set; }
    }

    public class TestRule : Rule
    {
        public override void Define()
        {
            IEnumerable<Order> orders = null;

            When()
                .Collect<Order>(() => orders, o => o.Cancelled);

            Then()
                .Do(ctx => orders.ToList().ForEach(o => Console.WriteLine("Matched order")));
        }
    }

    public class Program
    {
        static void Main(string[] args)
        {
            var repository = new RuleRepository();
            repository.Load(x => x.From(Assembly.GetExecutingAssembly()));

            var factory = repository.Compile();
            var session = factory.CreateSession();

            session.Insert(new Order(){Cancelled = true});
            session.Insert(new Order(){Cancelled = true});

            session.Fire();
        }
    }
```

**Comment by snikolayev** [comment]
@elgerm Regarding documentation. Have you seen the following section in wiki?
https://github.com/NRules/NRules/wiki/Fluent-Rules-DSL#rules-on-aggregates

**Comment by elgerm** [comment]
Yes I have, what's not clear to me is how NRules knows how to collect
these. In the program you add order1, order2 etc. And then collect them
with the collect method and put them in a collection.

Basically what I want is to add new orders at runtime:

            //Insert facts into rules engine's memory
            session.Insert(customer);
            session.Insert(order1);
            session.Insert(order2);
            //Start match/resolve/act cycle
            session.Fire();

And then, sometime later, another order comes in, I'd like to do this:

           var order3 = new Order(123458, customer, 2, 20.0);
           session.Insert(order3);
           session.Fire();

But the collect method doesn't fire for the third order. So I was wondering
how this works.

I then tried another thing and that was passing a collection of orders
every time and then fire, but I'm not sure if that's the way to go.

**Comment by gimmemoore** [comment]
HI @snikolayev!

Let me take a look at this over the weekend and come back to you, I was probably using an older version on .net 4.5.1 on Windows 7

Great work btw!

**Comment by snikolayev** [comment]
@elgerm Let me explain how it works. Remember I said that you don't want to insert the whole collection into the rules engine, and instead just want to insert individual facts. The reason is that behind the scenes NRules creates that collection for you as a synthetic fact. The rules that use the "Collect" pattern are activated when that synthetic fact is created, and then they fire when you call "Fire" method.
When you insert another fact after the rule has already fired, the synthetic collection is simply updated, by adding yet another fact to it. But the rule is already activated and has already fired, and so this update does not change the state of the rule.
This is not only true about synthetic facts. Same is true about regular facts. If you have a rule that matches on a Order, and that rule already matched and fired; if you update the Order and the rule still matches and you call "Fire" again - nothing will happen.
This is also how most other inference rules engines work. When a rule fires it may insert new facts that activate more rules, and duplicate fires on updates will break forward chaining.
This is actually the second time someone brings this up. I'm wrestling with a question whether I should add some sort of configuration to the engine to allow activating rules on updates. Still thinking about it....

For the specific thing you are trying to do, if you want to re-fire all rules every time you call "Fire" method, you can use a trick.
Create a special type of fact, lets call it Trigger.
```c#
public class Trigger {}
```
Then add this to every rule that you want to re-fire in a way I described:
```c#
When()
   .Match<Trigger>()
   //rest of patterns
```

Now, every time before you call "Fire" insert a new instance of the Trigger - it will create a new match for each rule that uses it, and for those rules that use "Collect" it will re-evaluate collection aggregation.
```c#
session.Insert(new Order());
session.Insert(new Trigger());
session.Fire();
//Collect captures one order
//...
session.Insert(new Order());
session.Insert(new Trigger());
session.Fire();
//Collect captures two orders
```

**Comment by snikolayev** [comment]
@elgerm I did some more analysis, and experimented with the rules engine using the scenario where collections are getting updated, and I think it's a bug that the rule does not get re-activated. It should work the way you intuitively expected.
I opened a new issue for that - #43 
I'll release new version with the fix in the next few days (I already committed the change to develop branch).

**Comment by elgerm** [comment]
Hi Sergiy,

Thank you!

About allowing the engine to activate the rules on updates, I got a use
case for you that would maybe shed some new light on this:

I'm writing software for a quality control project. It's about having a
coil of iron for packaging that can have many defects, there are sensors
that measure various defects as the coil is unrolled, processed and rolled
up again. During the process one would like to see the defects immediately
so they can for instance stop production. Same goes for the rules that are
being broken, you want to know immediately, not until the end of a
production cycle. A defect could be: a small hole in the coil. A rule could
be: 3 small holes near the side of the coil.

**Comment by snikolayev** [comment]
@elgerm Thanks, this is actually very helpful. I think you have a couple of options on how to implement this in NRules now. It also lines up nicely with some future features you can expect in the upcoming releases.
As I mentioned, right now there is no way to specify conditions on the aggregates (i.e. if you use "Collect", you cannot specify that you only want the rule to fire if Count == 3).
Option 1. Make collection a property on another fact (i.e. create a fact "Coil" that has a collection property Coil.Defects and have a rule ```Match<Coil>(c => c.Defects.Count == 3)```). Then you can add defects to that collection and call session.Update on the coil fact.
Option 2. Create a synthetic collection fact yourself (i.e. DefectSet) and insert/update that (kind of similar to option 1, except that you are creating a new fact and not modifying some existing fact).
Option 3. Check for collection count in the actions portion of the rule (I would avoid this if possible). 

Future considerations
I'm making changes as we speak to support conditions on the aggregates. I already committed most of the changes, just need to extend fluent DSL.

One of the major future features will be complex event processing. This will allow hooking event streams into the rules engine, where rules will fire as events occur. I think this is what you are looking for in terms of "not waiting until the end of production cycle". Until then, you need to call Fire explicitly when you want the engine to fire active rules.

**Comment by snikolayev** [comment]
FYI, release [0.3.1](https://github.com/NRules/NRules/releases/tag/v0.3.1) is out, which, among other things, adds support for conditions on aggregates, as well as re-activation of rules on updates.

**Comment by elgerm** [comment]
Thanks ill try it tomorrow

On Monday, February 9, 2015, Sergiy Nikolayev <notifications@github.com>
wrote:

> FYI, release 0.3.1 <https://github.com/NRules/NRules/releases/tag/v0.3.1>
> is out, which, among other things, adds support for conditions on
> aggregates, as well as re-activation of rules on updates.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NRules/NRules/issues/42#issuecomment-73459414>.
>


-- 
Met vriendelijke groet / Best regards,

Elger Mensonides
Deevelop B.V.
T: +31616965906
E: elger@deevelop.com
W: www.deevelop.com
